### PR TITLE
Fix PromotionIndividualCodeRedeemer non redeemed codes and unnecessar…

### PR DESCRIPTION
**Pull request by Drumm3r**
### 1. Why is this change necessary?

It is currently possible that individual codes will not be redeemed if there is more than one promotion in an Order. 
Additionally, there are many unnecessary database requests fired.

### 2. What does this change do, exactly?

It resolves this issue in PromotionIndividualCodeRedeemer and reduces the number of database requests that are fired.

### 3. Describe each step to reproduce the issue or behaviour.

1. Add at least one promotion with an individual code and one without to your cart. 
2. Order it. 
3. In some cases, it depends on the random order of the line items, the individual code will not be redeemed

### 4. Please link to the relevant issues (if any).

n/a

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a changelog file (adr/2023-07-31-fix-promotion-individual-code-redeemer.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### <samp>🤖 Generated by Copilot at 070ebb7</samp>

This pull request fixes a performance and reliability issue with the `PromotionIndividualCodeRedeemer` component that handles the redemption of individual promotion codes in the checkout process. It also adds an ADR document that explains the rationale and design of the solution.

### <samp>🤖 Generated by Copilot at 070ebb7</samp>

*  Add a new ADR document to explain the decision for fixing the promotion individual code redeemer issue (`adr/2023-07-31-fix-promotion-individual-code-redeemer.md`)
*  Avoid unnecessary database queries by checking if the code is null or empty before fetching the individual code entity ([link](https://github.com/shopware/platform/pull/3242/files?diff=unified&w=0#diff-e9b6d158353c513a2e67fe02d8ff6a3a086af263ae16b868ee14dae306902216R56-R60))
*  Ensure all the individual codes in the order are processed and redeemed by using continue instead of return when the individual code entity is not found or is invalid ([link](https://github.com/shopware/platform/pull/3242/files?diff=unified&w=0#diff-e9b6d158353c513a2e67fe02d8ff6a3a086af263ae16b868ee14dae306902216L67-R72))

### CODE REVIEW

File: adr/2023-07-31-fix-promotion-individual-code-redeemer.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+---
+title: Fix promotion individual code redeemer
```
Comment (shyim) [inline adr/2023-07-31-fix-promotion-individual-code-redeemer.md]: Can you add a changelog instead of an ADR. This is kinda just a bugfix :)

Comment (Drumm3r) [reply adr/2023-07-31-fix-promotion-individual-code-redeemer.md]: Of course. I changed it.

File: src/Core/Checkout/Promotion/Subscriber/PromotionIndividualCodeRedeemer.php [hunk 2]
```diff
@@ -64,7 +69,7 @@ public function onOrderPlaced(CheckoutOrderPlacedEvent $event): void
             // if we did not use an individual code we might have
             // just used a global one or anything else, so just quit in this case.
             if (!($individualCode instanceof PromotionIndividualCodeEntity)) {
-                return;
```
Comment (shyim) [inline src/Core/Checkout/Promotion/Subscriber/PromotionIndividualCodeRedeemer.php:67]: Could you add a small unit test for this case? So we don't miss this anymore in the future.

**Comment by bschulzebaek** [comment]
Hey @Drumm3r , I'm closing this PR due to inactivity - feel free to reopen it anytime. In that case, please review the requested changes 🙂

**Pull request closed without merging**